akka-persistence-typed/src/main/scala/akka/persistence/typed/keyvalue/internal/KeyValueBehaviorImpl.scala

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (C) 2017-2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.typed.keyvalue.internal
+
+import org.slf4j.LoggerFactory
+
+import akka.actor.typed
+import akka.actor.typed.ActorRef
+import akka.actor.typed.BackoffSupervisorStrategy
+import akka.actor.typed.Behavior
+import akka.actor.typed.BehaviorInterceptor
+import akka.actor.typed.PostStop
+import akka.actor.typed.Signal
+import akka.actor.typed.SupervisorStrategy
+import akka.actor.typed.internal.ActorContextImpl
+import akka.actor.typed.scaladsl.ActorContext
+import akka.actor.typed.scaladsl.Behaviors
+import akka.annotation._
+import akka.persistence.RecoveryPermitter
+import akka.persistence.typed.keyvalue.scaladsl._
+import akka.persistence.keyvaluestore.scaladsl.GetObjectResult
+import akka.persistence.typed.PersistenceId
+import akka.persistence.typed.SnapshotAdapter
+import akka.util.unused
+
+@InternalApi
+private[akka] object KeyValueBehaviorImpl {
+
+  /**
+   * Used by KeyValueBehaviorTestKit to retrieve the `persistenceId`.
+   */
+  final case class GetPersistenceId(replyTo: ActorRef[PersistenceId]) extends Signal
+
+  /**
+   * Used by KeyValueBehaviorTestKit to retrieve the state.
+   * Can't be a Signal because those are not stashed.
+   */
+  final case class GetState[State](replyTo: ActorRef[State]) extends InternalProtocol
+
+}
+
+@InternalApi
+private[akka] final case class KeyValueBehaviorImpl[Command, State](
```

Comment:
This is the starting point for the behavior implementation, with the 3 phases RequestingRecoveryPermit, Recovering and Running. The machinery is very similar to the event sourced implementation.